app/src/main/java/org/mozilla/fenix/components/toolbar/TabCounterToolbarButton.kt

```diff
@@ -8,34 +8,40 @@ import android.content.Context
 import android.util.TypedValue
 import android.view.View
 import android.view.ViewGroup
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.ExperimentalCoroutinesApi
 import mozilla.components.browser.menu.BrowserMenu
 import mozilla.components.browser.menu.BrowserMenuBuilder
 import mozilla.components.browser.menu.item.BrowserMenuDivider
 import mozilla.components.browser.menu.item.BrowserMenuImageText
-import mozilla.components.browser.session.Session
-import mozilla.components.browser.session.SessionManager
+import mozilla.components.browser.state.selector.normalTabs
+import mozilla.components.browser.state.selector.privateTabs
+import mozilla.components.browser.state.state.BrowserState
 import mozilla.components.concept.toolbar.Toolbar
+import mozilla.components.lib.state.ext.consumeFrom
 import org.mozilla.fenix.R
 import org.mozilla.fenix.components.metrics.Event
 import org.mozilla.fenix.ext.components
-import org.mozilla.fenix.ext.sessionsOfType
 import org.mozilla.fenix.ext.settings
 import org.mozilla.fenix.theme.ThemeManager
 import java.lang.ref.WeakReference
 
 /**
  * A [Toolbar.Action] implementation that shows a [TabCounter].
  */
+@ExperimentalCoroutinesApi
 class TabCounterToolbarButton(
-    private val sessionManager: SessionManager,
+    private val lifecycleOwner: LifecycleOwner,
     private val isPrivate: Boolean,
     private val onItemTapped: (TabCounterMenuItem) -> Unit = {},
     private val showTabs: () -> Unit
 ) : Toolbar.Action {
     private var reference: WeakReference<TabCounter> = WeakReference<TabCounter>(null)
 
     override fun createView(parent: ViewGroup): View {
-        sessionManager.register(sessionManagerObserver, view = parent)
+        parent.consumeFrom(parent.context.components.core.store, lifecycleOwner) {
+            updateCount(it)
```

Comment:
We should add an `FLow.ifChanged` call here to only update when the count changes